strawberry/types/type_resolver.py

```diff
@@ -222,36 +226,129 @@ def _resolve_types(fields: List[FieldDefinition]) -> List[FieldDefinition]:
+        A field defined using strawberry.field as a function, with a supplied
+        resolver.
+
+        The type hint is optional, but if supplied, it must match the return
+        type of the resolver. Type annnotation is required if resolver is not
+        type annotated; if both are annotated, they must match.
+
+        Implementation note: If both type annotations are provided, there will
+        be a redundant Type #1-style entry in the dataclass' field list. If the
+        strawberry.field call does not specify a `name`, the name of the field
+        on the class will be used.
+
+    Type #2c:
+        A field defined using @strawberry.field as a decorator around the
+        resolver. The resolver must be type-annotated.
+
+    Final `name` attribute priority:
+    1. Type #2 `name` attribute. This will be defined with an explicit
+       strawberry.field(name=...). No camelcase-ification will be done, as the
+       user has explicitly stated the field name.
+    2. Field name on the cls. Will exist for all fields other than Type #2c.
+       Field names will be converted to camelcase.
+
+    """
+    field_definitions: Dict[str, FieldDefinition] = {}
+
+    # Type #1 fields
+    type_1_fields: Dict[str, dataclasses.Field] = {
+        field.name: field for field in dataclasses.fields(cls)
+    }
+
+    # Type #2 fields
+    type_2_fields = {}
+    for field_name, field in cls.__dict__.items():
+        if hasattr(field, "_field_definition"):
+            type_2_fields[field_name] = field
+
+    for field_name, field in type_2_fields.items():
+
+        field_definition: FieldDefinition = field._field_definition
+
+        # Check if there is a matching Type #1 field:
+        if field_name in type_1_fields:
+            # Make sure field and resolver types are the same if both are
+            # defined
+            # TODO: Implement
+            # assert field.type == resolver.type
+
+            # Grab the type from the field if the resolver has no type
+            if field_definition.type is None:
+                field_type = type_1_fields[field_name].type
+                field_definition.type = field_type
+
+            # Stop tracking the Type #1 field, an explict strawberry.field was
+            # defined
+            type_1_fields.pop(field_name)
+
+        # Otherwise, ensure that a resolver has been specified
+        else:
+            if field_definition.base_resolver is None:
+                # This should be caught by _wrap_dataclass in type.py, but just
+                # in case, we'll check again
+                raise MissingResolverError(field_name)
 
-    # get all the fields from the dataclass
-    dataclass_fields = dataclasses.fields(cls)
+            # resolver with @strawberry.field decorator must be typed
+            if field_definition.type is None:
+                resolver_name = field_definition.base_resolver.__name__
+                raise MissingReturnAnnotationError(resolver_name)
```

Comment:
This check, and the one below, are currently swallowed by graphql.GraphQLSchema, which just returns a simple TypeError with our error's message.

I wonder if it would be best/possible to attempt calling `type_resolver._get_fields` _before_ building the schema?

Reply:
oh really? that's annoying. I think maybe we can move the check when we run the decorator?

Reply:
Well the issue is that `strawberry.field` doesn't know if it's being used as a decorator, or just with `resolver` as the only argument

As far as the `strawberry.field` can tell, these are equivalent, but only the first is valid:

```python
@strawberry.type
class Query:
    def my_func(): ...
    my_field: int = strawberry.field(resolver=my_func)

    @strawberry.field
    def my_func():
        ...
```

It doesn't know if the typing will/will not be provided externally on the field